### PR TITLE
Test on Pydantic v1 and v2

### PR DESCRIPTION
**Pull request by mattwthompson**
## Description
This PR updates the codebase to work with both Pydantic v1 and v2 and also tests each case.

### CODE REVIEW

**Review by j-wags — APPROVED** [review]
LGTM pending test success! The CI matrix is huge now, can we drop the pydantic v1 tests before merging?

**Comment by mattwthompson** [comment]
I'd prefer one of the following options:
1. Drop v1 from the test matrix and make a release based off of this that is explicitly `pydantic =2`
2. Keep just one v1 test in the matrix and make a release that stills supports v1 and v2

Does either sound acceptable to you? My preference is to do 2, since BespokeFit and some other downstreams don't yet work on v2, but follow this release up with one that drops v1 once it can

**Comment by mattwthompson** [comment]
I'm electing to go for option 2 as that allows #164 to slip in to environments that aren't yet updated for Pydantic v2

**Pull request merged**
